File: palindrome.js
function isPalindromeRecursive(str) {
  // Clean the string: convert to lowercase and remove non-alphanumeric characters
  const cleanedStr = str.toLowerCase().replace(/[^a-z0-9]/g, '');

  // Base case: if the string has 0 or 1 character, it's a palindrome
  if (cleanedStr.length <= 1) {
    return true;
  }

  // Recursive step: check if the first and last characters are the same
  // and if the substring between them is also a palindrome
  if (cleanedStr[0] === cleanedStr[cleanedStr.length - 1]) {
    return isPalindromeRecursive(cleanedStr.substring(1, cleanedStr.length - 1));
  } else {
    return false;
  }
}

// Example Usage (for verification)
console.log("Is 'madam' a palindrome?", isPalindromeRecursive('madam')); // true
console.log("Is 'A man, a plan, a canal: Panama' a palindrome?", isPalindromeRecursive('A man, a plan, a canal: Panama')); // true
console.log("Is 'hello' a palindrome?", isPalindromeRecursive('hello')); // false
console.log("Is 'racecar' a palindrome?", isPalindromeRecursive('racecar')); // true
console.log("Is 'No lemon, no melon' a palindrome?", isPalindromeRecursive('No lemon, no melon')); // true

module.exports = isPalindromeRecursive;

File: palindrome.py
#!/usr/bin/env python3

def is_palindrome_recursive(s):
    """
    Checks if a string is a palindrome using recursion.
    Ignores non-alphanumeric characters and case.
    """
    processed_s = ''.join(filter(str.isalnum, s)).lower()
    
    if len(processed_s) <= 1:
        return True
    else:
        return processed_s[0] == processed_s[-1] and is_palindrome_recursive(processed_s[1:-1])

if __name__ == '__main__':
    test_strings = [
        "madam",
        "A man, a plan, a canal: Panama",
        "racecar",
        "hello",
        "",
        "a",
        "No lemon, no melon"
    ]

    print("\n--- Palindrome Check Results ---")
    for s in test_strings:
        result = is_palindrome_recursive(s)
        print(f'"{s}" is a palindrome: {result}')
    print("--------------------------------")

